File: lee algo.py
from collections import deque
TOTAL_ROW = 12
TOTAL_COL = 10

class Cell:
	def __init__(self,x_cordinates: int, y_cordinates: int):
		self.x_cordinates = x_cordinates
		self.y_cordinates = y_cordinates

class node_queue:
	def __init__(self,pt: Cell, dist: int):
		self.pt = pt
		self.dist = dist

def check_row_and_col_is_valid_or_not(r: int, c: int):
	return ((r >= 0) and (r < TOTAL_ROW) and (c >= 0) and (c < TOTAL_COL))

r = [-1, 0, 0, 1]
c = [0, -1, 1, 0]

def lee_algo_function(matrix, source_point: Cell, destination_point: Cell):

	if matrix[source_point.x_cordinates][source_point.y_cordinates]!=1 or matrix[destination_point.x_cordinates][destination_point.y_cordinates]!=1:
		return -1

	visited_cell = [[False for x in range(TOTAL_COL)]
					for y in range(TOTAL_ROW)]

	visited_cell[source_point.x_cordinates][source_point.y_cordinates] = True

	queue = deque()

	ce = node_queue(source_point,0)
	queue.append(ce)

	while queue:

		current = queue.popleft()

		pt = current.pt
		if pt.x_cordinates == destination_point.x_cordinates and pt.y_cordinates == destination_point.y_cordinates:
			return current.dist

		for x in range(4):
			row = pt.x_cordinates + r[x]
			col = pt.y_cordinates + c[x]

			if (check_row_and_col_is_valid_or_not(row,col) and
			matrix[row][col] == 1 and
				not visited_cell[row][col]):
				visited_cell[row][col] = True
				adjency_Cell = node_queue(Cell(row,col),
									current.dist+1)
				queue.append(adjency_Cell)
	return -1

matrix = 	[
		[ 1, 0, 1, 1, 1, 1, 0, 1, 1, 1 ],
		[ 1, 0, 1, 0, 1, 1, 1, 0, 1, 1 ],
		[ 1, 1, 1, 0, 1, 1, 0, 1, 0, 1 ],
		[ 0, 0, 0, 0, 1, 0, 0, 0, 0, 1 ],
		[ 1, 1, 1, 0, 1, 1, 1, 0, 1, 0 ],
		[ 1, 0, 1, 1, 1, 1, 0, 1, 0, 0 ],
		[ 1, 0, 0, 0, 0, 0, 0, 0, 0, 1 ],
		[ 1, 0, 1, 1, 1, 1, 0, 1, 1, 1 ],
        [ 1, 0, 1, 1, 1, 1, 0, 1, 0, 0 ],
		[ 1, 0, 0, 0, 0, 0, 0, 0, 0, 1 ],
		[ 1, 0, 1, 1, 1, 1, 0, 1, 1, 1 ],
		[ 1, 1, 0, 0, 0, 0, 1, 0, 0, 1 ]
	];
source_point = Cell(0,0)
destination_point = Cell(5,4)

distance = lee_algo_function(matrix,source_point,destination_point)

if distance!=-1:
		print("Length of the Shortest Path is: ",distance)
else:
		print("Shortest Path doesn't exist")

File: lee algo.cpp
#include <bits/stdc++.h>
using namespace std;
#define TOTAL_ROW 12
#define TOTAL_COL 10
struct Point
{
	int x_cordinates;
	int y_cordinates;
};
struct node_queue
{
	Point pt;
	int distance;
};
bool check_row_and_col_is_valid_or_not(int r, int c)//here r used for Row and c used for col
{
	return (r >= 0) && (r < TOTAL_ROW) &&
		(c >= 0) && (c < TOTAL_COL);
}
int r[] = {-1, 0, 0, 1};
int c[] = {0, -1, 1, 0};
int lee_algo_function(int matrix[][TOTAL_COL], Point source_point, Point destination_point)
{
	if (!matrix[source_point.x_cordinates][source_point.y_cordinates] || !matrix[destination_point.x_cordinates][destination_point.y_cordinates])
		return -1;

	bool visited_cell[TOTAL_ROW][TOTAL_COL];
	memset(visited_cell, false, sizeof visited_cell);
	visited_cell[source_point.x_cordinates][source_point.y_cordinates] = true;
	queue<node_queue> queue;
	node_queue ce = {source_point, 0};
	queue.push(ce);
	while (!queue.empty())
	{
		node_queue current = queue.front();
		Point pt = current.pt;
		if (pt.x_cordinates == destination_point.x_cordinates && pt.y_cordinates == destination_point.y_cordinates)
			return current.distance;
		queue.pop();
		for (int x = 0; x < 4; x++)
		{
			int row = pt.x_cordinates + r[x];
			int col = pt.y_cordinates + c[x];
			if (check_row_and_col_is_valid_or_not(row, col) && matrix[row][col] && !visited_cell[row][col])
			{
				visited_cell[row][col] = true;
				node_queue adjency_Cell = { {row, col},current.distance + 1 };
				queue.push(adjency_Cell);
			}
		}
	}
	return -1;
}
int main()
{
	int matrix[TOTAL_ROW][TOTAL_COL] =
	{
		{ 1, 0, 1, 1, 1, 1, 0, 1, 1, 1 },
		{ 1, 0, 1, 0, 1, 1, 1, 0, 1, 1 },
		{ 1, 1, 1, 0, 1, 1, 0, 1, 0, 1 },
		{ 0, 0, 0, 0, 1, 0, 0, 0, 0, 1 },
		{ 1, 1, 1, 0, 1, 1, 1, 0, 1, 0 },
		{ 1, 0, 1, 1, 1, 1, 0, 1, 0, 0 },
		{ 1, 0, 0, 0, 0, 0, 0, 0, 0, 1 },
		{ 1, 0, 1, 1, 1, 1, 0, 1, 1, 1 },
        { 1, 0, 1, 1, 1, 1, 0, 1, 0, 0 },
		{ 1, 0, 0, 0, 0, 0, 0, 0, 0, 1 },
		{ 1, 0, 1, 1, 1, 1, 0, 1, 1, 1 },
		{ 1, 1, 0, 0, 0, 0, 1, 0, 0, 1 }
	};

	Point source_point = {0, 0};
	Point destination_point = {5, 4};

	int distance = lee_algo_function(matrix, source_point, destination_point);

	if (distance != -1)
		cout <<"Shortest Path is: " << distance ;
	else
		cout <<"Shortest Path doesn't exist";

	return 0;
}
